File: backend/app/schemas/final_project.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.project_submission import ProjectSubmissionStatus


class ProjectRequirementCreate(BaseModel):
    description: str
    is_mandatory: bool = True
    order: int = 0


class ProjectRequirementResponse(ProjectRequirementCreate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int


class TrainingProjectCreate(BaseModel):
    title: str
    description: str | None = None
    passing_score: int = 75
    is_active: bool = True
    requirements: list[ProjectRequirementCreate] = Field(default_factory=list)


class TrainingProjectUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    passing_score: int | None = None
    is_active: bool | None = None
    requirements: list[ProjectRequirementCreate] | None = None


class ProjectReviewCreate(BaseModel):
    score: int
    feedback: str


class ProjectReviewResponse(ProjectReviewCreate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    submission_id: int
    reviewer_id: int
    created_at: datetime
    updated_at: datetime


class ProjectSubmissionCreate(BaseModel):
    github_url: str | None = None
    live_url: str | None = None
    file_url: str | None = None
    student_notes: str | None = None


class ProjectSubmissionUpdate(ProjectSubmissionCreate):
    status: Literal["SUBMITTED"] | None = None


class ProjectSubmissionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    student_id: int
    github_url: str | None
    live_url: str | None
    file_url: str | None
    student_notes: str | None
    status: ProjectSubmissionStatus
    submitted_at: datetime | None
    created_at: datetime
    updated_at: datetime
    reviews: list[ProjectReviewResponse] = Field(default_factory=list)


class TrainingProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    track_id: int
    title: str
    description: str | None
    passing_score: int
    is_active: bool
    created_at: datetime
    updated_at: datetime
    requirements: list[ProjectRequirementResponse] = Field(default_factory=list)

File: frontend/src/types/finalProject.ts
export type ProjectSubmissionStatus =
  | "DRAFT"
  | "SUBMITTED"
  | "UNDER_REVIEW"
  | "CHANGES_REQUIRED"
  | "APPROVED"
  | "REJECTED";

export interface ProjectRequirement {
  id: number;
  project_id: number;
  description: string;
  is_mandatory: boolean;
  order: number;
}

export interface TrainingProject {
  id: number;
  track_id: number;
  title: string;
  description: string | null;
  passing_score: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  requirements: ProjectRequirement[];
}

export interface ProjectReview {
  id: number;
  submission_id: number;
  reviewer_id: number;
  score: number | null;
  feedback: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProjectSubmission {
  id: number;
  project_id: number;
  student_id: number;
  github_url: string | null;
  live_url: string | null;
  file_url: string | null;
  student_notes: string | null;
  status: ProjectSubmissionStatus;
  submitted_at: string | null;
  created_at: string;
  updated_at: string;
  reviews: ProjectReview[];
}

export interface ProjectSubmissionCreate {
  github_url?: string | null;
  live_url?: string | null;
  file_url?: string | null;
  student_notes?: string | null;
}

export interface ProjectSubmissionUpdate extends ProjectSubmissionCreate {
  status?: "SUBMITTED";
}

export interface ProjectReviewCreate {
  score: number;
  feedback: string;
}
